tests/app/Infrastructure/Controller/PostBuyCoinControllerTest.php

```diff
@@ -4,70 +4,89 @@
-        $this->walletDataSource = Mockery::mock(WalletDataSource::class);
         $this->app->bind(CoinDataSource::class, function () {
             return $this->coinDataSource;
         });
-        $this->app->bind(WalletDataSource::class, function () {
-            return $this->walletDataSource;
-        });
     }
 
     /**
      * @test
      */
-    public function ifCoinIdNotFoundThrowsError()
+    public function doesNotBuyCoinWhenCoinDoesNotExist()
     {
         $this->coinDataSource
             ->expects("findById")
             ->with("coin_id_value", "1")
             ->andReturn(null);
 
-        $response = $this->post('api/coin/buy', ["coin_id" => "coin_id_value",
-                                                      "wallet_id" => "wallet_id_value",
-                                                      "amount_usd" => 1]);
+        $response = $this->post(
+            'api/coin/buy',
+            [
+                "coin_id" => "coin_id_value",
+                "wallet_id" => "wallet_id_value",
+                "amount_usd" => 1
+            ]
+        );
 
-        $response->assertNotFound();
-        $response->assertExactJson(['description' => 'A coin with the specified ID was not found.']);
+        $response->assertStatus(Response::HTTP_NOT_FOUND);
+        $response->assertExactJson(
+            [
+                'description' => 'A coin with the specified ID was not found.'
+            ]
+        );
     }
 
     /**
      * @test
      */
-    public function ifBadRequestThrowsError()
+    public function doesNotBuyCoinWhenWalletDoesNotExist()
     {
         $this->coinDataSource
             ->expects("findById")
             ->with("coin_id_value", "1")
-            ->times(0)
-            ->andReturn(null);
+            ->andReturn(new Coin(
+                "coin_id_value",
+                "name_value",
+                "symbol_value",
+                1,
+                1
+            ));
+        Cache::shouldReceive('has')->once()->with('wallet_0')->andReturn(false);
 
-        $response = $this->post('api/coin/buy', ["coin_id" => null,
-                                                    "wallet_id" => null,
-                                                    "amount_usd" => null]);
+        $response = $this->post(
+            'api/coin/buy',
+            [
+                "coin_id" => "coin_id_value",
+                "wallet_id" => "0",
+                "amount_usd" => 1
+            ]
+        );
 
-        $response->assertBadRequest();
-        $response->assertExactJson(['description' => 'bad request error']);
+        $response->assertStatus(Response::HTTP_NOT_FOUND);
+        $response->assertExactJson(
+            [
+                'description' => 'A wallet with the specified ID was not found'
+            ]
+        );
     }
 
     /**
      * @test
      */
-    public function ifCoinWasBoughtCorrectlyReturnSuccessful()
+    public function buysCoinWhenCoinAndWalletExist()
```

Comment:
You don't always have to put why.
There are some cases like the successfull ones that it is clearer if `successfullBuyOperation`